pgxn/neon/extension_server.c

```diff
@@ -77,6 +79,62 @@ neon_download_extension_file_http(const char *filename, bool is_library)
 	return ret;
 }
 
+static void
+DownloadRemoteExtensionOnUpdate(
+	PlannedStmt *pstmt,
+	const char *queryString,
+	bool readOnlyTree,
+	ProcessUtilityContext context,
+	ParamListInfo params,
+	QueryEnvironment *queryEnv,
+	DestReceiver *dest,
+	QueryCompletion *qc)
+{
+	Node	   *parseTree = pstmt->utilityStmt;
+
+	switch (nodeTag(parseTree))
+	{
+		case T_AlterExtensionStmt:
+		{
+			AlterExtensionStmt *stmt = castNode(AlterExtensionStmt, parseTree);
+
+			/*
+			 * Unconditionally try to download a remote extension when updating
+			 */
+			neon_download_extension_file_http(stmt->extname, false);
```

Comment:
I don't necessarily love this solution since there will be a WARNING visible to users on most `ALTER EXTENSION UPDATE` statements, but this is the best I could think of without creating a commit in the Postgres fork, which sounds like a worse solution to me, but I'm happy to do if that's what people think is best.

Reply:
It is handled using a hook in the fork. See the hook in this file we are talking about. There is something to be said about handling that in the same way, but obviously we would accrue additional latency when installing an extension. I'm always motivated by reducing the fork diff if possible.

Reply:
I don't remember, it's not possible to print WARNING only to logs, but not to client, right?

@lubennikovaav unresolved thread to hear your opinion re this change. Generally, I think the best flow is to only issue remote extensions call if we cannot find some files locally. IIUC works this way when Postgres preloads libraries, but not for CREATE EXTENSION. Can we follow the same semantics everywhere?

Reply:
@tristan957 Let's downgrade this WARNING to LOG - we only need it to debug issues with remote extensions and we can easily find it, no matter what message level is used.

>I think the best flow is to only issue remote extensions call if we cannot find some files locally.

+1, I think I initially implemented it two keep the fork diff minimal and to allow remote extensions to overwrite local ones, but we never needed this flow in reality.